janusgraph-es/src/main/java/org/janusgraph/diskstorage/es/rest/RestElasticSearchClient.java

```diff
@@ -426,6 +427,58 @@ public long countTotal(String indexName, Map<String, Object> requestData) throws
         }
     }
 
+    private double executeAggs(String indexName, Map<String, Object> requestData, String agg, String fieldName) throws IOException {
```

Comment:
I think we should also add documentation about aggregation loose of precision for big numbers when using ElasticSearch. As noted here, for any number more then `2^53` the result is approximate: https://www.elastic.co/guide/en/elasticsearch/reference/current/search-aggregations.html#limits-for-long-values

I think users should now about this limitation and if they want precise values they need to disable this optimization in their queries. I.e.:
```
g.traversal().withoutStrategies(JanusGraphMixedIndexAggStrategy.class).has("someProperty", 123).sum()
```

Reply:
I'm also thinking that maybe it would make sense to actually keep Count aggregation strategy separate from other aggregations optimization strategy. If so, users might want to opt-in using and accept aggregation limits it or not using it and to not have any limits.

Reply:
> I'm also thinking that maybe it would make sense to actually keep Count aggregation strategy separate from other aggregations optimization strategy

That makes sense because even ES count (without aggregation) itself is an approximate...

Reply:
> That makes sense because even ES count (without aggregation) itself is an approximate...

I think you are referring to the old implementation (probably in 0.2 or 0.3 version of JanusGraph) where we were using [total hits](https://www.elastic.co/guide/en/elasticsearch/reference/7.17/search-your-data.html#track-total-hits).

Right now for ElasticSearch implementation we are using [Count API](https://www.elastic.co/guide/en/elasticsearch/reference/current/search-count.html) which suppose to return accurate value unless I'm missing something. At least I didn't find information that it's not accurate.

So, my point is that users most likely want to have `count` optimization in their queries all the times because it's accurate but aggregation optimization isn't accurate when the resulting value is more then 2^53. So, some users may want to exclude aggregation optimization in some cases but have count optimization enabled in their queries.

That's why I suggested to have both optimizations for `count` and for `aggregation`.

Another reason is that:
`g.V().has("foo", "bar").limit(1000).count()` - can use optimization with a specific logic like `Math.min(limit, countResultFromIndex)` but I can't say the same for other aggregation queries.

Moreover, later (not necessary in this PR) I think we will need to add a configuration parameter which excludes this aggregation optimization by default - otherwise it may be a breaking change for some users due to approximate results in some cases.

Reply:
> Right now for ElasticSearch implementation we are using [Count API](https://www.elastic.co/guide/en/elasticsearch/reference/current/search-count.html) which suppose to return accurate value unless I'm missing something. At least I didn't find information that it's not accurate.

You are right! Never mind, I misremembered it :)

Reply:
ElasticSearch count API is accurate but the Lucene one is not. It only counts the 1000 first elements ([TotalHits](https://lucene.apache.org/core/8_11_2/core/org/apache/lucene/search/TotalHits.html)). It was not the case in Lucene 7.7.2 (JanusGraph 5.x).

I agree with you, accurate optimizations should not be mixed with approximate one. I can split it.